src/man/man3/StrChooser.3
.TH StringChooser 3I "30 November 1989" "InterViews" "InterViews Reference Manual"
.SH NAME
StringChooser \- dialog box for choosing items in a list
.SH SYNOPSIS
.B #include <InterViews/strchooser.h>
.SH DESCRIPTION
StringChooser is a dialog that manages keyboard focus between a
StringBrowser and a StringEditor to provide a convenient interface to
selecting strings.  The user can select strings either by selecting
them in the StringBrowser or by typing them in the StringEditor. Like
Dialog, StringChooser is an abstract class; programmers must derive
from StringChooser to add push buttons, etc. and to compose them into
custom arrangements.
.PP
The user can focus keyboard input towards either the StringEditor or
the StringBrowser.  To set focus to the StringEditor, for example, the
user left-clicks in its canvas to make it interpret subsequent
keyboard events.  When either the StringEditor or the StringBrowser
relinquishes keyboard focus, the other acquires it: for example, the
user can toggle focus between the two by pressing the tab key
repeatedly.  If an entry in the StringBrowser is selected when it
loses focus, the (first) selected string will be transferred into the
StringEditor.
.SH PUBLIC OPERATIONS
.TP
.B "StringChooser(ButtonState*, int rows, int cols, const char* sample, \
Alignment)"
.ns
.TP
.B "StringChooser(const char* name, ButtonState*, int rows, int cols, \
const char* sample, Alignment)"
.ns
Construct a new StringChooser with the given button state and pop-up
alignment.  These constructors create a StringBrowser of the specified
size and a StringEditor with the given sample string.  The
StringChooser, StringBrowser, and StringEditor all share the same
button state.
.TP
.B "void Select(int point)"
.ns
.TP
.B "void Select(int left, int right)"
.ns
.TP
.B "void SelectMessage()"
Select an insertion point, a subrange, or the entire edit buffer in
the StringEditor.
.TP
.B "void Message(const char* text)"
Set the contents of the StringEditor's edit buffer to \fItext\fP.
.TP
.B "virtual const char* Choice()"
Return the chosen string. The string is returned in a static buffer
and should be copied before use.
.SH PROTECTED OPERATIONS
.TP
.B "StringChooser(ButtonState*, Alignment = Center)"
Construct a new StringChooser with the given button state and pop-up
alignment \fIwithout\fP creating StringBrowser and StringEditor
instances.  This constructor should be called in subclasses that
require a specialized StringEditor or StringBrowser or both.
.TP
.B "void Init(StringEditor*, StringBrowser*)"
Initialize the StringEditor and StringBrowser that StringChooser will
use.  This function is useful in the constructors of StringChooser
subclasses that require a specialized StringEditor or StringBrowser or
both.
.TP
.B "virtual void SwitchFocus()"
Prepare to shift keyboard focus to the StringBrowser if the
StringEditor is focused currently, and vice versa.  Focus actually
changes when HandleFocus is called (see below).
.TP
.B "virtual boolean CanFocus(Interactor*)"
Return whether or not the given interactor can be focused.  This
function is used to determine whether the StringChooser should switch
focus in response to a downclick.
.TP
.B "virtual void HandleFocus()"
Transfer event reading to the StringBrowser or StringEditor, whichever
is the current focus.
.TP
.B "virtual void UpdateEditor()"
.ns
.TP
.B "virtual void UpdateBrowser()"
Update StringEditor and StringBrowser state when the dialog is
accepted or in response to a change in focus.  This function
guarantees that the Choice operation will return the last choice. By
default, UpdateEditor changes the StringEditor's contents to match the
StringBrowser selection (if any), and UpdateBrowser clears out any
selected strings in the StringBrowser.
.SH SEE ALSO
Button(3I), Dialog(3I), StringBrowser(3I), StringEditor(3I)


CHANGES
May 12th 2012  ivtools-1.2.10

Many practical and small extensions to comterp, the scripting language embedded in
comdraw, flipbook, graphdraw, and drawserv.

Dropped complete support for gcc 3.0 and earlier. Things still might work but you'd be better off
upgradeing to gcc 4.0 or staying with ivtools-1.2.9.

Feb. 23rd, 2010  ivtools-1.2.9

- Debian related changes suggested by Agustin Martin.

- usec timeout on comdraw pause command -- allows for writing loops
  that can be exited.

- option for an end-of-line token introduced to LexScan.

- port to 64 bit Fedora

- fix problem with negative numbers in attribute editor

- make File/Save and File/Revert always work regardless of modified flag

Nov. 21st, 2009  ivtools-1.2.8

- significant work on keeping up with gcc and libstdc++.  A wrapper
  for stdio.h was inadvertently getting pulled in by /usr/include
  files, causing a lot of trouble.  Now builds with gcc-4.4 on Ubuntu
  9.10.

August 25th, 2009  ivtools-1.2.7

- series of memory leak checking commands: alistleak, attrvleak, commandleak, compleak, graphlicleak, mlineleak, viewleak.  Each returns the current number of allocated objects of that type (AttributeValueList, AttributeValue, Command, OverlayComp, Graphic, MultiLineObj, and OverlayView).

- added reference counting to OverlayView's (ComponentView's) referred from AttributeValue's (#define RESOURCE_COMPVIEW)

- fixed attrname func.

- fixed stddev func.

- added BooleanType to NumFunc promotion method.

- fixed graphdraw copy/paste of graph fragments

- new SocketObj for use comterp

- fix behavior of symbol manipulating funcs, by using internal bquote.

- port to Ubuntu Heron

- applied (most of) patch to build on Fedora 10 contributed by John Heidemann

- balance of changes to support ipl-1.0.4

July 9th, 2008  ivtools-1.2.6

- Debian contributed patch for iostreams backward compatibility

- bug fix for dispatcher.c contributed by Damon Permezel

- add mute command to comterp

- balance of changes to support ipl-1.0.2

March 4th, 2008  ivtools-1.2.5

- overall evolution to support use by ipl-1.0.1 (see
  http://sf.net/projects/ipl for further details)

July 20th, 2006  ivtools-1.2.4

- changes for compiling with gcc-4.0

- changes to socket-driven (ACE) use of comterp to support multi-line commands

- changes to allow re-use of scanner/parser with multiple operator tables

October 7th, 2005  ivtools-1.2.3

- add support for an external libtiff with a --with-tiff (and
-with-tiff-lib) configure arguments.

- changes necessary to compile with gcc-4.0

- changes necessary to compile with gcc-4.0 on top of ACE-5.4.7

- changes necessary to compile on suse-9.3 linux (amd64 platform).
Contributed by Paul Vixie.

- fixed bug in "addtool" command/mechanism in graphdraw and drawserv

- added support for color expressions to graphdraw (:colexpr), to
allow attribute-sensitive computation of a node's color.

January 25th, 2005  ivtools-1.2.2

If you submitted a patch in the past half year to ivtools that you
don't see reflected here, please resubmit.  I lost that e-mail.

- fix for problem writing to /tmp

- compilation fix on NetBSD for comterp.c

- compilation patch submitted by Michal Palczew.

July 13th, 2004  ivtools-1.2.1

Two crashes discovered and fixed by Michal Palczewski (mpalczew at
users.sourceforge.net).  One was a crash on exit after a delete with
idraw/drawtool, the other was a crash upon delete with graphdraw.
They turned out to be long-standing bugs in the order of object
deletion, problems that hadn't appeared with other versions of gcc.
RedHat must have adopted a more aggressive recycling of freed space in
a recent release.  Thanks Michal.


June 3rd, 2004  ivtools-1.2

ivtools-1.2 sets a new direction for the package, toward peer-to-peer
drawing chat.  ivtools' drawserv, the editor that incorporates every ivtool
mechanism to date (multi-frame, graph-editing, integrated scripting
language) has been extended to support a unique kind of
small-granularity shared whiteboard.  

This work was based on the pre-existing networked scripting capability
of ivtools' comdraw.  Then the pasting of any graphic was intercepted,
and the graphic serialized into a command that could be communicated
to other drawing editors to keep them in sync.

Then a connection management system was established, that supported
two-way peer-to-peer links between drawing editors.  In addition the
return path on each socket was used for detecting timeout, which
causes the connection to be dropped.

Finally, a distributed mutually-exclusive graphic selection mechanism
was created, where every drawing editor is aware of who currently has
a graphic selected or who had it selected last if it is not currently
selected.  When one drawing editor tries to select a graphic, it first
requests that permission from the last selector (after determining it
is unselected).

Now you can link to other drawing editors, draw graphics and see them
appear remotely (although only in black and white), and exercise the
distributed selection mechanism.  Four pieces remain to be developed:
1) support for full-up graphic state transmittal (it is sent, but not
interpreted on the remote end);, 2) support for transmission of the
complete set of drawing editor commands; 3) extending the distributed
mutual-exclusion mechanism to select pseudo-graphics to acquire global
permission for connection, pasting, moving front-to-back, and
grouping; and 4) a two-tier mechanism for invoking select(), that will
allow the bypass of selection requests while engaged in lengthy direct
manipulations.

If all this interests you, and you'd like to help, build this version
of ivtools with ACE support, subscribe to the ivtools-dev mailing, and
help me do some testing, programming, thinking, whatever.  Or visit
http://www.ivtools.org to make a donation to support this work.  Thanks.

Scott Johnston




src/IVGlyph/strchooser.c
/*
 * Copyright (c) 1993 David B. Hollenbeck
 *
 * Permission to use, copy, modify, distribute, and sell this software and 
 * its documentation for any purpose is hereby granted without fee, provided
 * that (i) the above copyright notice and this permission notice appear in
 * all copies of the software and related documentation, and (ii) the name of
 * David B. Hollenbeck may not be used in any advertising or
 * publicity relating to the software without the specific, prior written
 * permission of David B. Hollenbeck.
 * 
 * THE SOFTWARE IS PROVIDED "AS-IS" AND WITHOUT WARRANTY OF ANY KIND, 
 * EXPRESS, IMPLIED OR OTHERWISE, INCLUDING WITHOUT LIMITATION, ANY 
 * WARRANTY OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  
 *
 * IN NO EVENT SHALL DAVID B. HOLLENBECK BE LIABLE FOR
 * ANY SPECIAL, INCIDENTAL, INDIRECT OR CONSEQUENTIAL DAMAGES OF ANY KIND,
 * OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
 * WHETHER OR NOT ADVISED OF THE POSSIBILITY OF DAMAGE, AND ON ANY THEORY OF 
 * LIABILITY, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE 
 * OF THIS SOFTWARE.
 */
/*
 * Copyright (c) 1991 Stanford University
 * Copyright (c) 1991 Silicon Graphics, Inc.
 *
 * Permission to use, copy, modify, distribute, and sell this software and 
 * its documentation for any purpose is hereby granted without fee, provided
 * that (i) the above copyright notices and this permission notice appear in
 * all copies of the software and related documentation, and (ii) the names of
 * Stanford and Silicon Graphics may not be used in any advertising or
 * publicity relating to the software without the specific, prior written
 * permission of Stanford and Silicon Graphics.
 * 
 * THE SOFTWARE IS PROVIDED "AS-IS" AND WITHOUT WARRANTY OF ANY KIND, 
 * EXPRESS, IMPLIED OR OTHERWISE, INCLUDING WITHOUT LIMITATION, ANY 
 * WARRANTY OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  
 *
 * IN NO EVENT SHALL STANFORD OR SILICON GRAPHICS BE LIABLE FOR
 * ANY SPECIAL, INCIDENTAL, INDIRECT OR CONSEQUENTIAL DAMAGES OF ANY KIND,
 * OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
 * WHETHER OR NOT ADVISED OF THE POSSIBILITY OF DAMAGE, AND ON ANY THEORY OF 
 * LIABILITY, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE 
 * OF THIS SOFTWARE.
 */

/*
 * StrChooser -- select a string
 */

#include <IV-look/choice.h>
#include "strchooser.h"
#include <IV-look/dialogs.h>
#include <IV-look/fbrowser.h>
#include <IV-look/kit.h>
#include <InterViews/action.h>
#include <InterViews/font.h>
#include <InterViews/hit.h>
#include <InterViews/layout.h>
#include <InterViews/style.h>
#include <InterViews/target.h>
#include <OS/string.h>

class StrChooserImpl {
protected:
    strchooser_callback outfunc_;
    void* base_;

private:
    friend class StrChooser;

    WidgetKit  * kit_;
    StrChooser * fchooser_;
    FileBrowser* fbrowser_;
    StrChooserAction* action_;
    GlyphIndex   selected_;
    Style      * style_;
    Action     * update_;
    StringList * strings_;
    String     * caption_;
    boolean      embed_;

    void init(StrChooser*, Style*, StrChooserAction*);
    void free();
    void build();
    void clear();
    void accept_browser();
    void cancel_browser();
};

declareActionCallback(StrChooserImpl)
implementActionCallback(StrChooserImpl)

StrChooser::StrChooser(StringList* sl, String* caption, WidgetKit* kit, Style* s, 
    StrChooserAction* a, boolean embed, strchooser_callback outfunc, void* base) 
: Dialog(nil, s) {
    impl_ = new StrChooserImpl;
    StrChooserImpl& sci = *impl_;
    sci.kit_ = kit;
    sci.strings_ = sl;
    sci.caption_ = caption;
    sci.embed_ = embed;
    sci.outfunc_ = outfunc;
    sci.base_ = base;
    sci.init(this, s, a);
}

StrChooser::~StrChooser() {
    impl_->free();
    delete impl_;
}

GlyphIndex StrChooser::selected() const {
    return impl_->selected_;
}

void StrChooser::dismiss(boolean accept) {
    Dialog::dismiss(accept);
    StrChooserImpl& fc = *impl_;
    if (fc.action_ != nil) {
	fc.action_->execute(this, accept);
    }
}

/** class StrChooserImpl **/

void StrChooserImpl::init(
    StrChooser* chooser, Style* s, StrChooserAction* a
) {
    fchooser_ = chooser;
    fbrowser_ = nil;
    Resource::ref(a);
    action_ = a;
    style_ = new Style(s);
    Resource::ref(style_);
    style_->alias("StrChooser");
    style_->alias("Dialog");
    update_ = new ActionCallback(StrChooserImpl)(
	this, &StrChooserImpl::build
    );
    style_->add_trigger_any(update_);
    build();
}

void StrChooserImpl::free() {
    Resource::unref(action_);
    style_->remove_trigger_any(update_);
    Resource::unref(style_);
}

void StrChooserImpl::build() {
    WidgetKit& kit = *kit_;
    const LayoutKit& layout = *LayoutKit::instance();
    Style* s = style_;
    kit.push_style();
    kit.style(s);
    String caption("");
    s->find_attribute("caption", caption);
    s->find_attribute("subcaption", *caption_);
    String open("OK");
    s->find_attribute("ok", open);
    String close("Cancel");
    s->find_attribute("cancel", close);
    long rows = 10;
    s->find_attribute("rows", rows);
    const Font* f = kit.font();
    FontBoundingBox bbox;
    f->font_bbox(bbox);
    Coord height = rows * (bbox.ascent() + bbox.descent()) + 1.0;
    Coord width;
    if (!s->find_attribute("width", width)) {
	Coord defwidth = 16 * f->width('m') + 3.0;
	Coord maxwidth = 0;
	for (ListItr(StringList) i(*strings_); i.more(); i.next()) {
	  Coord curwidth = f->width(i.cur().string(), i.cur().length()) + 3.0;
	  if (curwidth > maxwidth)
	    maxwidth = curwidth;
	}
	width = maxwidth > defwidth ? maxwidth : defwidth;
    }

    Action* accept = new ActionCallback(StrChooserImpl)(
	this, &StrChooserImpl::accept_browser
    );
    Action* cancel = new ActionCallback(StrChooserImpl)(
	this, &StrChooserImpl::cancel_browser
    );
    fbrowser_ = new FileBrowser(kit_, accept, cancel);

    fchooser_->remove_all_input_handlers();
    fchooser_->append_input_handler(fbrowser_);

    Glyph* g = layout.vbox();
    if (caption.length() > 0) {
	g->append(layout.rmargin(kit.fancy_label(caption), 5.0, fil, 0.0));
    }
    if (caption_->length() > 0) {
	g->append(layout.rmargin(kit.fancy_label(*caption_), 5.0, fil, 0.0));
    }
    g->append(layout.vglue(5.0, 0.0, 2.0));
    g->append(
	layout.hbox(
	    layout.vcenter(
		kit.inset_frame(
		    layout.margin(
			layout.natural_span(fbrowser_, width, height), 1.0
		    )
		),
		1.0
	    ),
	    layout.hspace(4.0),
	    kit.vscroll_bar(fbrowser_->adjustable())
	)
    );
    if (!embed_) {
        g->append(layout.vspace(15.0));
    	g->append(
	    layout.hbox(
	    	layout.hglue(10.0),
	    	layout.vcenter(kit.default_button(open, accept)),
	    	layout.hglue(10.0, 0.0, 5.0),
	    	layout.vcenter(kit.push_button(close, cancel)),
	    	layout.hglue(10.0)
	    )
    	);

        fchooser_->body(
	    layout.back(
	        layout.vcenter(kit.outset_frame(layout.margin(g, 5.0)), 1.0),
	        new Target(nil, TargetPrimitiveHit)
	    )
        );
    } else {
        fchooser_->body(
	    layout.back(
	        layout.vcenter(g, 1.0), new Target(nil, TargetPrimitiveHit)
	    )
        );
    }
    fchooser_->focus(fbrowser_);

    for (ListItr(StringList) i(*strings_); i.more(); i.next()) {
        Glyph* name = kit.label(i.cur());
        Glyph* label = new Target(
	    layout.h_margin(name, 3.0, 0.0, 0.0, 15.0, fil, 0.0),
    	    TargetPrimitiveHit);
        TelltaleState* t = new TelltaleState(TelltaleState::is_enabled);
        fbrowser_->append_selectable(t);
        fbrowser_->append(new ChoiceItem(t, label, kit.bright_inset_frame(label)));
        fbrowser_->refresh();
    }

    kit.pop_style();
}

void StrChooserImpl::clear() {
    Browser& b = *fbrowser_;
    b.select(-1);
    GlyphIndex n = b.count();
    for (GlyphIndex i = 0; i < n; i++) {
	b.remove_selectable(0);
	b.remove(0);
    }
}

void StrChooserImpl::accept_browser() {
    GlyphIndex i = fbrowser_->selected();
    if (i == -1) {
	return;
    }
    selected_ = i;
    if (embed_ && outfunc_ != nil)
       (*outfunc_)(base_);

    fchooser_->dismiss(true);
}

void StrChooserImpl::cancel_browser() {
    selected_ = -1;
    fchooser_->dismiss(false);
}


/** class StrChooserAction **/

StrChooserAction::StrChooserAction() { }
StrChooserAction::~StrChooserAction() { }
void StrChooserAction::execute(StrChooser*, boolean) { }
